### PR TITLE
Fix the issue that we cannot modify `apm_config.obfuscation.*` through env vars

### PR DESCRIPTION
**Pull request by keisku**
### What does this PR do?

Fix the issue that we cannot modify `apm_config.obfuscation.*` through env vars.

Set default values based on this file.

https://github.com/DataDog/datadog-agent/blob/26052f9b0fdb75103002f7090cab64a7c9a12aa2/pkg/config/config_template.yaml#L1159-L1267

### Motivation

There’s no special meaning behind choosing `apm_config.obfuscatio.memcached.enabled` and `apm_config.obfuscatio.http.remove_paths_with_digits`. They’re just examples.

For example, when we try to change `apm_config.obfuscatio.memcached.enabled` and `apm_config.obfuscatio.http.remove_paths_with_digits` in container environment, we would try to set env vars like this.

```bash
docker run --name datadog-agent -d --cgroupns host --pid host \
-v /var/run/docker.sock:/var/run/docker.sock:ro \
-v /proc/:/host/proc/:ro -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
-e DD_API_KEY=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx \
-e DD_APM_OBFUSCATION_MEMCACHED_ENABLED=false \
-e DD_APM_OBFUSCATION_HTTP_REMOVE_PATHS_WITH_DIGITS=false \
datadog/agent:7.60.0

docker exec -it datadog-agent agent config
apm_config:
  # ...skip...
  obfuscation.memcached.enabled: "false"
  obfuscation.http.remove_paths_with_digits: "false"
```

The following outputs are not expected. They should be hierarchical structure.
- `obfuscation.memcached.enabled: "false"`
- `obfuscation.http.remove_paths_with_digits: "false"`

Workaround

Change configs through a static file.

```bash
docker run --name datadog-agent -d --cgroupns host --pid host \
-v /var/run/docker.sock:/var/run/docker.sock:ro \
-v /proc/:/host/proc/:ro -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
-e DD_API_KEY=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx \
$(docker build -t datadog-agent --quiet -<<-EOD
FROM datadog/agent:7.60.0
COPY --chmod=755 <<"CONFIGEND" /etc/datadog-agent/datadog.yaml
apm_config:
  obfuscation:
    memcached:
      enabled: false
    http:
      remove_paths_with_digits: false
CONFIGEND
EOD
)

docker exec -it datadog-agent agent config
apm_config:
  # ...skip...
  obfuscation:
    memcached:
      enabled: false
    http:
      remove_paths_with_digits: false
```

### Describe how you validated your changes

```bash
docker run --name datadog-agent -d --cgroupns host --pid host \
-v /var/run/docker.sock:/var/run/docker.sock:ro \
-v /proc/:/host/proc/:ro -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
-e DD_API_KEY=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx \
datadog/agent-dev:keisku-apm-config-obfuscation-py3

docker exec -it datadog-agent agent config
apm_config:
  # The following configs are default.
  obfuscation:
    cache:
      enabled: true
    elasticsearch:
      enabled: true
      keep_values: []
      obfuscate_sql_values: []
    http:
      remove_paths_with_digits: false
      remove_query_string: false
    memcached:
      enabled: true
      keep_command: false
    mongodb:
      enabled: true
      keep_values: []
      obfuscate_sql_values: []
    opensearch:
      enabled: true
      keep_values: []
      obfuscate_sql_values: []
    redis:
      enabled: true
      remove_all_args: true
    remove_stack_traces: false
    sql_exec_plan:
      enabled: false
      keep_values: []
      obfuscate_sql_values: []
    sql_exec_plan_normalize:
      enabled: false
      keep_values: []
      obfuscate_sql_values: []

docker run --name datadog-agent -d --cgroupns host --pid host \
-v /var/run/docker.sock:/var/run/docker.sock:ro \
-v /proc/:/host/proc/:ro -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
-e DD_API_KEY=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx \
-e DD_APM_OBFUSCATION_MEMCACHED_ENABLED=false \
-e DD_APM_OBFUSCATION_HTTP_REMOVE_PATHS_WITH_DIGITS=false \
-e DD_APM_OBFUSCATION_SQL_EXEC_PLAN_NORMALIZE_KEEP_VALUES='value1 value2 value3' \
datadog/agent-dev:keisku-apm-config-obfuscation-py3

docker exec -it datadog-agent agent config
apm_config:
  # ...skip....
  obfuscation:
    cache:
      enabled: true
    elasticsearch:
      enabled: true
      keep_values: []
      obfuscate_sql_values: []
    http:
      remove_paths_with_digits: false
      remove_query_string: false
    memcached:
      # MODIFIED
      enabled: false
      keep_command: false
    mongodb:
      enabled: true
      keep_values: []
      obfuscate_sql_values: []
    opensearch:
      enabled: true
      keep_values: []
      obfuscate_sql_values: []
    redis:
      enabled: true
      remove_all_args: true
    remove_stack_traces: false
    sql_exec_plan:
      enabled: false
      keep_values: []
      obfuscate_sql_values: []
    sql_exec_plan_normalize:
      enabled: false
      # MODIFIED
      keep_values:
      - value1
      - value2
      - value3
      obfuscate_sql_values: []
```

### CODE REVIEW

**Review by pgimalac — COMMENTED** [review]
I don't think I understand what the issue was or how your changes fix it, can you add some details in the PR description ?

**Review by rtrieu — APPROVED** [review]

**Comment by keisku** [comment]
@pgimalac Thanks for reviewing this PR. I put the detail here.

> I don't think I understand what the issue was or how your changes fix it, can you add some details in the PR description ?

The goal is to solve the issue that we cannot change `apm_config.obfuscation.*` through env vars.

For example, when we try to change `apm_config.obfuscatio.memcached.enabled` and `apm_config.obfuscatio.http.remove_paths_with_digits` in container environment, we would try to set env vars like this.
There’s no special meaning behind choosing `apm_config.obfuscatio.memcached.enabled` and `apm_config.obfuscatio.http.remove_paths_with_digits`. They’re just examples.

```bash
docker run --name datadog-agent -d --cgroupns host --pid host \
-v /var/run/docker.sock:/var/run/docker.sock:ro \
-v /proc/:/host/proc/:ro -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
-e DD_API_KEY=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx \
-e DD_APM_OBFUSCATION_MEMCACHED_ENABLED=false \
-e DD_APM_OBFUSCATION_HTTP_REMOVE_PATHS_WITH_DIGITS=false \
datadog/agent:7.60.0
```

Then, we would check if configs are reflected by env vars like `docker exec -it datadog-agent agent config`.

We can expect output like this.

```yaml
apm_config:
  obfuscation:
    memcached:
      enabled: false
    http:
      remove_paths_with_digits: false
```

BUT, the actual output is like this.

```yaml
apm_config:
  obfuscation.memcached.enabled: "false"
  obfuscation.http.remove_paths_with_digits: "false"
```

`obfuscation.memcached.enabled` and `obfuscation.http.remove_paths_with_digits` are expected to have a hierarchical structure, but it does not. Besides, the value is string, not boolean.

To solve this issue, I opened this PR.

**Review by pgimalac — APPROVED** [review]

**Comment by pgimalac** [comment]
Thanks for the PR, we'll look into the bug in the config

File: pkg/collector/python/datadog_agent.go [hunk 1]
```diff
@@ -264,8 +264,11 @@ func lazyInitObfuscator() *obfuscate.Obfuscator {
 		if !cfg.SQLExecPlanNormalize.Enabled {
 			cfg.SQLExecPlanNormalize = defaultSQLPlanNormalizeSettings
 		}
-		if !cfg.Mongo.Enabled {
-			cfg.Mongo = defaultMongoObfuscateSettings
+		if len(cfg.Mongo.KeepValues) == 0 {
+			cfg.Mongo.KeepValues = defaultMongoObfuscateSettings.KeepValues
+		}
+		if len(cfg.Mongo.ObfuscateSQLValues) == 0 {
+			cfg.Mongo.ObfuscateSQLValues = defaultMongoObfuscateSettings.ObfuscateSQLValues
```
Comment (keisku) [inline pkg/collector/python/datadog_agent.go:271]: Before this PR, `cfg` here is not reflected by default values because `DD_APM_OBFUSCATION_*` are not bound to default values. This is the issue that I want to solve with this PR.

After this PR, `cfg.Mongo.Enabled` is `true` since `DD_APM_OBFUSCATION_MONGODB_ENABLED` is bound to `true`. This follows this documentation.

https://github.com/DataDog/datadog-agent/blob/62c67dbe190ebb1128ce7091b1459e2105d896e9/pkg/config/config_template.yaml#L1212-L1215

So, we need this change to keep backward compatibility.

`cfg.SQLExecPlan.Enabled` and `cfg.SQLExecPlanNormalize.Enabled` are `false` by default because `DD_APM_OBFUSCATION_SQL_EXEC_PLAN_ENABLED` and `DD_APM_OBFUSCATION_SQL_EXEC_PLAN_NORMALIZE_ENABLED` are bound to `false`. Therefore, we can keep backward compatibility without code change here.

File: pkg/config/setup/apm.go [hunk 2]
```diff
@@ -21,29 +21,29 @@ import (
...
+	config.BindEnvAndSetDefault("apm_config.obfuscation.sql_exec_plan_normalize.obfuscate_sql_values", []string{}, "DD_APM_OBFUSCATION_SQL_EXEC_PLAN_NORMALIZE_OBFUSCATE_SQL_VALUES")
+	config.BindEnvAndSetDefault("apm_config.obfuscation.http.remove_query_string", false, "DD_APM_OBFUSCATION_HTTP_REMOVE_QUERY_STRING")
+	config.BindEnvAndSetDefault("apm_config.obfuscation.http.remove_paths_with_digits", false, "DD_APM_OBFUSCATION_HTTP_REMOVE_PATHS_WITH_DIGITS")
+	config.BindEnvAndSetDefault("apm_config.obfuscation.remove_stack_traces", false, "DD_APM_OBFUSCATION_REMOVE_STACK_TRACES")
+	config.BindEnvAndSetDefault("apm_config.obfuscation.redis.enabled", true, "DD_APM_OBFUSCATION_REDIS_ENABLED")
+	config.BindEnvAndSetDefault("apm_config.obfuscation.redis.remove_all_args", false, "DD_APM_OBFUSCATION_REDIS_REMOVE_ALL_ARGS")
+	config.BindEnvAndSetDefault("apm_config.obfuscation.memcached.enabled", true, "DD_APM_OBFUSCATION_MEMCACHED_ENABLED")
+	config.BindEnvAndSetDefault("apm_config.obfuscation.memcached.keep_command", false, "DD_APM_OBFUSCATION_MEMCACHED_KEEP_COMMAND")
+	config.BindEnvAndSetDefault("apm_config.obfuscation.cache.enabled", true, "DD_APM_OBFUSCATION_CACHE_ENABLED")
```
[the remark this answers is not in the document]
Comment (keisku) [reply pkg/config/setup/apm.go:46]: https://github.com/DataDog/datadog-agent/pull/32318/commits/8b1b9d6193e7dec6985a9402e259c5571cd34add is for this comment.

**Comment by keisku** [comment]
@ajgajg1134

Thank you for reviewing this PR! I am documenting the testing method and results here. I will add the Unit Test later.

Background:
Let me explain the background of this PR more.
#31336 introduced obfuscation caching to address the CPU usage issue. I wanted to test the cache feature but noticed that it isn’t enabled for the core-agent, and I can’t enable it through an environment variable.

For investigation, I added debug logs that dump obfuscation config to the code before this PR (= main branch).

TL;DR: We can modify obfuscation config through env vars for trace-agent, but not core-agent.

<details><summary>git diff</summary>

```diff
git diff HEAD~1 | cat -
diff --git a/comp/trace/config/setup.go b/comp/trace/config/setup.go
index 06a157dc64..0c26271855 100644
--- a/comp/trace/config/setup.go
+++ b/comp/trace/config/setup.go
@@ -508,6 +508,7 @@ func applyDatadogConfig(c *config.AgentConfig, core corecompcfg.Component) error
 		if pkgconfigsetup.Datadog().IsSet("apm_config.obfuscation.cache.enabled") {
 			c.Obfuscation.Cache.Enabled = pkgconfigsetup.Datadog().GetBool("apm_config.obfuscation.cache.enabled")
 		}
+		log.Debugf("obfuscation config: %#v", c.Obfuscation)
 	}
 
 	if core.IsSet("apm_config.filter_tags.require") {
diff --git a/pkg/collector/python/datadog_agent.go b/pkg/collector/python/datadog_agent.go
index 95a0b2f69f..af02378d79 100644
--- a/pkg/collector/python/datadog_agent.go
+++ b/pkg/collector/python/datadog_agent.go
@@ -258,6 +258,7 @@ func lazyInitObfuscator() *obfuscate.Obfuscator {
 			log.Errorf("Failed to unmarshal apm_config.obfuscation: %s", err.Error())
 			cfg = obfuscate.Config{}
 		}
+		log.Debugf("obfuscation config: %#v", cfg)
 		if !cfg.SQLExecPlan.Enabled {
 			cfg.SQLExecPlan = defaultSQLPlanObfuscateSettings
 		}
diff --git a/pkg/trace/agent/obfuscate.go b/pkg/trace/agent/obfuscate.go
index 25042e3fd6..228c2433f3 100644
--- a/pkg/trace/agent/obfuscate.go
+++ b/pkg/trace/agent/obfuscate.go
@@ -138,6 +138,7 @@ func (a *Agent) lazyInitObfuscator() *obfuscate.Obfuscator {
 
 	if a.obfuscator == nil {
 		if a.obfuscatorConf != nil {
+			log.Debugf("obfuscation config: %#v", *a.obfuscatorConf)
 			a.obfuscator = obfuscate.NewObfuscator(*a.obfuscatorConf)
 		} else {
 			a.obfuscator = obfuscate.NewObfuscator(obfuscate.Config{})
```

</details>

Then, I checked debug logs with the following containers.

<details><summary>docker-compose.yaml</summary>

```yaml
services:
  agent:
    container_name: datadog-agent
    image: datadog/agent-dev:keisku-apm-config-obfuscation-investigation-py3
    volumes:
      - /sys/fs/cgroup/:/host/sys/fs/cgroup:ro
      - /proc/:/host/proc/:ro
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - /var/lib/cloud/data/instance-id:/var/lib/cloud/data/instance-id:ro
    environment:
      DD_LOG_LEVEL: debug
      DD_HOSTNAME_FILE: /var/lib/cloud/data/instance-id
      DD_APM_OBFUSCATION_MEMCACHED_ENABLED: false
      DD_APM_OBFUSCATION_CACHE_ENABLED: true
      DD_APM_OBFUSCATION_SQL_EXEC_PLAN_ENABLED: true
      DD_APM_OBFUSCATION_SQL_EXEC_PLAN_KEEP_VALUES: value1 value2 value3
    env_file:
      - ~/sandbox.docker.env
  mysql:
    container_name: mysql
    image: mysql:8
    restart: always
    labels:
      com.datadoghq.ad.checks: |
        {
          "mysql": {
            "instances": [
              {
                "host": "%%host%%",
                "port": "3306",
                "username": "datadog",
                "password": "datadog",
                "dbm": "true"
              }
            ]
          }
        }
    environment:
      MYSQL_ROOT_PASSWORD: password
      MYSQL_USER: datadog
      MYSQL_PASSWORD: datadog
    volumes:
    - ./init.sh:/docker-entrypoint-initdb.d/init.sh
```

</details>

We can modify obfuscation config through env vars for trace-agent, but not core-agent.

The result of this test case, I am sure env vars don't change obfuscation config for CORE.

- `DD_APM_OBFUSCATION_MEMCACHED_ENABLED: false` is reflected for TRACE and CORE both.
- `DD_APM_OBFUSCATION_CACHE_ENABLED: true` is reflected for TRACE, but not CORE.
  - TRACE | `Cache:obfuscate.CacheConfig{Enabled:true}`
  - CORE | `Cache:obfuscate.CacheConfig{Enabled:false}`, it should be `true`.
- `DD_APM_OBFUSCATION_SQL_EXEC_PLAN_ENABLED: true` and `DD_APM_OBFUSCATION_SQL_EXEC_PLAN_KEEP_VALUES: value1 value2 value3` are reflected for TRACE, but not CORE.
  - TRACE | `SQLExecPlan:obfuscate.JSONConfig{Enabled:true, KeepValues:[]string{"value1", "value2", "value3"}`
  - CORE | `SQLExecPlan:obfuscate.JSONConfig{Enabled:false, KeepValues:[]string(nil), ObfuscateSQLValues:[]string(nil)}`

```bash
docker logs datadog-agent | grep 'obfuscation\ config\:'

2024-12-21 09:49:38 UTC | TRACE | DEBUG | (comp/trace/config/setup.go:511 in applyDatadogConfig) | obfuscation config: &config.ObfuscationConfig{ES:obfuscate.JSONConfig{Enabled:true, KeepValues:[]string(nil), ObfuscateSQLValues:[]string(nil)}, OpenSearch:obfuscate.JSONConfig{Enabled:true, KeepValues:[]string(nil), ObfuscateSQLValues:[]string(nil)}, Mongo:obfuscate.JSONConfig{Enabled:true, KeepValues:[]string(nil), ObfuscateSQLValues:[]string(nil)}, SQLExecPlan:obfuscate.JSONConfig{Enabled:true, KeepValues:[]string{"value1", "value2", "value3"}, ObfuscateSQLValues:[]string(nil)}, SQLExecPlanNormalize:obfuscate.JSONConfig{Enabled:false, KeepValues:[]string(nil), ObfuscateSQLValues:[]string(nil)}, HTTP:obfuscate.HTTPConfig{RemoveQueryString:false, RemovePathDigits:false}, RemoveStackTraces:false, Redis:obfuscate.RedisConfig{Enabled:true, RemoveAllArgs:false}, Memcached:obfuscate.MemcachedConfig{Enabled:false, KeepCommand:false}, CreditCards:obfuscate.CreditCardsConfig{Enabled:true, Luhn:false, KeepValues:[]string(nil)}, Cache:obfuscate.CacheConfig{Enabled:true}}

2024-12-21 09:49:42 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:261 in func1) | obfuscation config: obfuscate.Config{SQL:obfuscate.SQLConfig{DBMS:"", TableNames:false, CollectCommands:false, CollectComments:false, CollectProcedures:false, ReplaceDigits:false, KeepSQLAlias:false, DollarQuotedFunc:false, ObfuscationMode:"", RemoveSpaceBetweenParentheses:false, KeepNull:false, KeepBoolean:false, KeepPositionalParameter:false, KeepTrailingSemicolon:false, KeepIdentifierQuotation:false, KeepJSONPath:false, Cache:false}, ES:obfuscate.JSONConfig{Enabled:false, KeepValues:[]string(nil), ObfuscateSQLValues:[]string(nil)}, OpenSearch:obfuscate.JSONConfig{Enabled:false, KeepValues:[]string(nil), ObfuscateSQLValues:[]string(nil)}, Mongo:obfuscate.JSONConfig{Enabled:false, KeepValues:[]string(nil), ObfuscateSQLValues:[]string(nil)}, SQLExecPlan:obfuscate.JSONConfig{Enabled:false, KeepValues:[]string(nil), ObfuscateSQLValues:[]string(nil)}, SQLExecPlanNormalize:obfuscate.JSONConfig{Enabled:false, KeepValues:[]string(nil), ObfuscateSQLValues:[]string(nil)}, HTTP:obfuscate.HTTPConfig{RemoveQueryString:false, RemovePathDigits:false}, Redis:obfuscate.RedisConfig{Enabled:false, RemoveAllArgs:false}, Memcached:obfuscate.MemcachedConfig{Enabled:false, KeepCommand:false}, CreditCard:obfuscate.CreditCardsConfig{Enabled:false, Luhn:false, KeepValues:[]string(nil)}, Statsd:obfuscate.StatsClient(nil), Logger:obfuscate.Logger(nil), Cache:obfuscate.CacheConfig{Enabled:false}}
```

**Review by ajgajg1134 — APPROVED** [review]
looks good, thanks for the added details and tests here!

**Comment by keisku** [comment]
> ERROR: Merge-base of this branch is too old for SMP. Please update your branch by merging an up-to-date main branch into your branch or by rebasing it on an up-to-date main branch.
https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/748647629 

I am going to update this branch.

File: pkg/collector/python/datadog_agent.go [hunk 3]
```diff
@@ -253,21 +254,23 @@ var (
...
 		}
-		if !cfg.SQLExecPlanNormalize.Enabled {
-			cfg.SQLExecPlanNormalize = defaultSQLPlanNormalizeSettings
+		if !obfuscaterConfig.SQLExecPlanNormalize.Enabled {
+			obfuscaterConfig.SQLExecPlanNormalize = defaultSQLPlanNormalizeSettings
 		}
-		if !cfg.Mongo.Enabled {
-			cfg.Mongo = defaultMongoObfuscateSettings
+		if len(obfuscaterConfig.Mongo.KeepValues) == 0 {
+			obfuscaterConfig.Mongo.KeepValues = defaultMongoObfuscateSettings.KeepValues
```
Comment (keisku) [inline pkg/collector/python/datadog_agent.go:268]: Before this PR, `cfg` here is not reflected by default values because `DD_APM_OBFUSCATION_*` are not bound to default values. This is the issue that I want to solve with this PR.

After this PR, `cfg.Mongo.Enabled` is `true` since `DD_APM_OBFUSCATION_MONGODB_ENABLED` is bound to `true`. This follows this documentation.

https://github.com/DataDog/datadog-agent/blob/62c67dbe190ebb1128ce7091b1459e2105d896e9/pkg/config/config_template.yaml#L1212-L1215

So, we need this change to keep backward compatibility.

`cfg.SQLExecPlan.Enabled` and `cfg.SQLExecPlanNormalize.Enabled` are `false` by default because `DD_APM_OBFUSCATION_SQL_EXEC_PLAN_ENABLED` and `DD_APM_OBFUSCATION_SQL_EXEC_PLAN_NORMALIZE_ENABLED` are bound to `false`. Therefore, we can keep backward compatibility without code change here.

**Review by vickenty — APPROVED** [review]
